codegen/src/main/kotlin/software/amazon/smithy/rust/codegen/smithy/protocols/parse/JsonParserGenerator.kt

```diff
@@ -0,0 +1,457 @@
+    }
+
+    private fun RustWriter.deserializeTimestamp(member: MemberShape) {
+        val timestampFormat =
+            httpBindingResolver.timestampFormat(
+                member, HttpLocation.DOCUMENT,
+                TimestampFormatTrait.Format.EPOCH_SECONDS
+            )
+        val timestampFormatType = RuntimeType.TimestampFormat(runtimeConfig, timestampFormat)
+        rustTemplate("#{expect_timestamp_or_null}(tokens.next(), #{T})?", "T" to timestampFormatType, *codegenScope)
+    }
+
+    private fun RustWriter.deserializeCollection(shape: CollectionShape) {
+        val fnName = symbolProvider.deserializeFunctionName(shape)
+        val parser = RuntimeType.forInlineFun(fnName, "json_deser") {
+            it.rustBlockTemplate(
+                """
+                pub fn $fnName<'a, I>(tokens: &mut #{Peekable}<I>) -> Result<Option<#{Shape}>, #{Error}>
+                    where I: Iterator<Item = Result<#{Token}<'a>, #{Error}>>
+                """,
+                "Shape" to symbolProvider.toSymbol(shape),
+                *codegenScope,
+            ) {
+                startArrayOrNull {
+                    rust("let mut items = Vec::new();")
+                    rustBlock("loop") {
+                        rustBlock("match tokens.peek()") {
+                            rustBlockTemplate("Some(Ok(#{Token}::EndArray { .. })) =>", *codegenScope) {
+                                rust("tokens.next().transpose().unwrap(); break;")
+                            }
+                            rustBlock("_ => ") {
+                                withBlock("items.push(", ");") {
+                                    deserializeMember(shape.member, false)
+                                }
+                            }
+                        }
+                    }
+                    rust("return Ok(Some(items))")
+                }
+            }
+        }
+        rust("#T(tokens)?", parser)
+    }
+
+    private fun RustWriter.deserializeMap(shape: MapShape) {
+        val keyTarget = model.expectShape(shape.key.target) as StringShape
+        val fnName = symbolProvider.deserializeFunctionName(shape)
+        val parser = RuntimeType.forInlineFun(fnName, "json_deser") {
+            it.rustBlockTemplate(
+                """
+                pub fn $fnName<'a, I>(tokens: &mut #{Peekable}<I>) -> Result<Option<#{Shape}>, #{Error}>
+                    where I: Iterator<Item = Result<#{Token}<'a>, #{Error}>>
+                """,
+                "Shape" to symbolProvider.toSymbol(shape),
+                *codegenScope,
+            ) {
+                startObjectOrNull {
+                    rust("let mut map = #T::new();", RustType.HashMap.RuntimeType)
+                    objectKeyLoop {
+                        withBlock("map.insert(", ");") {
+                            if (keyTarget.hasTrait<EnumTrait>()) {
+                                rust(
```

Comment:
this is probably worth extracting `parseStringInner` or something because of the enum-parsing logic, use of `escaped_str`, etc.

Reply:
I knew you would comment on this ;)

I had attempted to commonize the enum/string handling logic between `deserializeString` and `deserializeMap`, but they're just different enough that it doesn't really make a lot of sense. For example, `deserializeString` reads a token with different functions based on if its a string or enum, while `deserializeMap` already has a token with an `EscapedStr` in it and needs to decide what to do with that.

I guess I could make a common function that operates on an `EscapedStr` and then refactor `deserializeString` to pipe that into it. That may just work.